### PR TITLE
Enabled parameterized bucketing of event data

### PR DESCRIPTION
**Pull request by DuncanSmith1126**
Unit tests pending approval of feature

### CODE REVIEW

File: pkg/sloop/queries/queryd3heatmap.go [hunk 1]
```diff
@@ -63,11 +63,26 @@ func EventHeatMap3Query(params url.Values, t typed.Tables, queryStartTime time.T
...
+	eventBucketWidthString := params.Get(BucketingParam)
+	if eventBucketWidthString == "" {
+		eventBucketWidthString = "1m"
+	}
+
+	eventBucketWidth, err := time.ParseDuration(eventBucketWidthString)
+	if err != nil {
+		return []byte{}, fmt.Errorf("compute event bucket width failed with error: %v", err)
+	}
+	if eventBucketWidth < time.Minute {
```
[the remark this answers is not in the document]
Comment (DuncanSmith1126) [reply pkg/sloop/queries/queryd3heatmap.go:75]: Uhh - I'm OK with weird requests giving weird results if it helps keep the code a bit neater. I could enforce minute-by-minute granularity - or just have the results cut off if the entire minute isn't within the request window. The behavior is weird either way - considering how we bucket events.

File: pkg/sloop/webfiles/sloop_ui.js [hunk 2]
```diff
@@ -396,7 +396,7 @@ function createResourceBar(d) {
                 .attr("rx", 6)
                 .attr("ry", 6)
                 .attr("height", yAxisBand.bandwidth() * 0.6)
-                .attr("width", overlayW * 0.75)
+                .attr("width", overlayW * 0.9999) // Leave a tiny amount of wiggle room - chart looks goofy without it
```
[the remark this answers is not in the document]
Comment (DuncanSmith1126) [reply pkg/sloop/webfiles/sloop_ui.js:399]: There's a lot of UI magic in here - I'll try and write some more comments in after I'm sure this works front-end wise.

File: pkg/sloop/queries/queryd3heatmap.go [hunk 3]
```diff
@@ -63,11 +63,26 @@ func EventHeatMap3Query(params url.Values, t typed.Tables, queryStartTime time.T
...
+		return []byte{}, fmt.Errorf("event bucket width cannot be less than one minute")
+	}
+
 	// add the event counts in as overlay
-	mapResSumKeyToOverlay, err := eventCountsToOverlayMap(rawRows.Events)
+	var mapResSumKeyToOverlay map[typed.ResourceSummaryKey][]Overlay
+	mapResSumKeyToOverlay, err = eventCountsToOverlayMapBucketed(rawRows.Events, eventBucketWidth)
```
[the remark this answers is not in the document]
Comment (DuncanSmith1126) [reply pkg/sloop/queries/queryd3heatmap.go:81]: Reverted

File: pkg/sloop/queries/queryd3heatmap.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (DuncanSmith1126) [reply pkg/sloop/queries/queryd3heatmap.go:75]: Done

File: pkg/sloop/queries/queryd3heatmap.go [hunk 4]
```diff
@@ -63,11 +64,27 @@ func EventHeatMap3Query(params url.Values, t typed.Tables, queryStartTime time.T
 		return []byte{}, err
 	}
 
+	eventBucketWidthString := params.Get(BucketingParam)
+	if eventBucketWidthString == "" {
+		eventBucketWidthString = "1m"
+	}
+
+	eventBucketWidth, err := time.ParseDuration(eventBucketWidthString)
```
Comment (mallow111) [inline pkg/sloop/queries/queryd3heatmap.go:72]: what is the unite of eventBucketWidth here, should we force to be minutes, I noticed in the blow function we assume it is minute.

File: pkg/sloop/queries/queryd3heatmap.go [hunk 5]
```diff
@@ -63,11 +64,27 @@ func EventHeatMap3Query(params url.Values, t typed.Tables, queryStartTime time.T
...
+	if err != nil {
+		return []byte{}, fmt.Errorf("compute event bucket width failed with error: %v", err)
+	}
+
+	if eventBucketWidth < minBucketWidth {
```
Comment (mallow111) [inline pkg/sloop/queries/queryd3heatmap.go:77]: should we also control the upper bound?

File: pkg/sloop/queries/queryd3heatmap_test.go [hunk 6]
```diff
@@ -106,6 +106,7 @@ func helper_UrlValues() url.Values {
 	return map[string][]string{
 		NamespaceParam: []string{AllNamespaces},
 		KindParam:      []string{AllKinds},
+		BucketingParam: {"1m"},
```
Comment (mallow111) [inline pkg/sloop/queries/queryd3heatmap_test.go:109]: should we put this 1m as a constant somewhere, so it can be used/updated eaisier?

**Comment by nurland** [comment]
Stale PR

**Pull request closed without merging**